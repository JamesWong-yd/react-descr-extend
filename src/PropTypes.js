const reactDescApiFunctions = [
  'description',
  'deprecated',
  'format',
  'defaultProp',
];
const handler = {
  get: (target, prop) => {
    /* eslint-disable no-param-reassign */
    if (!target.reactDesc) {
      target.reactDesc = {};
    }
    if (reactDescApiFunctions.indexOf(prop) >= 0) {
      return (value) => {
        target.reactDesc[prop] = value;
        return new Proxy(target, handler);
      };
    } else if (prop === 'isRequired') {
      target.reactDesc.required = true;
      return new Proxy(target, handler);
    }
    /* eslint-enable no-param-reassign */
    return target[prop];
  },
};

const createProxyFunc = type => args => new Proxy(
  { type, args, isRequired: { type, args, required: true } }, handler,
);

export default new Proxy({
  any: { type: 'any', isRequired: { type: 'any', required: true } },
  array: { type: 'array', isRequired: { type: 'array', required: true } },
  bool: { type: 'bool', isRequired: { type: 'bool', required: true } },
  element: { type: 'element', isRequired: { type: 'element', required: true } },
  func: { type: 'func', isRequired: { type: 'func', required: true } },
  node: { type: 'node', isRequired: { type: 'node', required: true } },
  number: { type: 'number', isRequired: { type: 'number', required: true } },
  object: { type: 'object', isRequired: { type: 'object', required: true } },
  string: { type: 'string', isRequired: { type: 'string', required: true } },
  symbol: { type: 'symbol', isRequired: { type: 'symbol', required: true } },
  arrayOf: createProxyFunc('arrayOf'),
  custom: (callback) => {
    const target = callback.bind(null);
    target.type = 'func';
    return new Proxy(target, handler);
  },
  instanceOf: createProxyFunc('instanceOf'),
  objectOf: createProxyFunc('objectOf'),
  oneOfType: createProxyFunc('oneOfType'),
  oneOf: createProxyFunc('oneOf'),
  shape: createProxyFunc('shape'),
}, {
  get: (target, prop) => (
    new Proxy(typeof target[prop] === 'object' ? { ...target[prop] } : target[prop], handler)
  ),
});
