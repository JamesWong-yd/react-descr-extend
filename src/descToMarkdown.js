import descToJSON from './descToJSON';

export default function descToMarkdown(component, reactDesc) {
  if (!component) {
    throw new Error('react-desc: component is required');
  }

  const documentation = descToJSON(component, reactDesc);
  let description = '';
  if (documentation.description) {
    description = `\n${documentation.description}`;
  }
  let deprecatedNote = '';
  let deprecatedContent = '';
  if (documentation.deprecated) {
    deprecatedNote = '~~';
    deprecatedContent = `\n### Deprecated: ${documentation.deprecated}`;
  }
  let properties = '';
  if (documentation.properties) {
    properties = '\n\n### Properties\n\n| Property | Description | Format | Default Value | Required | Details |\n| ---- | ---- | ---- | ---- | ---- | ---- |';
    documentation.properties.forEach((prop) => {
      let propDeprecatedNote = '';
      let propDeprecatedContent = '';
      if (prop.deprecated) {
        propDeprecatedNote = '~~';
        propDeprecatedContent = `**Deprecated**: ${prop.deprecated}`;
      }
      properties += `\n| **${propDeprecatedNote}${prop.name}${propDeprecatedNote}** | ${prop.description} | ${prop.format.replace(/\n/g, '<br/>')} | ${prop.defaultValue || ''} | ${prop.required ? 'Yes' : 'No'} | ${propDeprecatedContent} |`;
    });
  }
  return `## ${deprecatedNote}${documentation.name}${deprecatedNote} Component${deprecatedContent}${description}${properties}`;
}
