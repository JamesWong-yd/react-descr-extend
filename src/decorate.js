import PropTypes from 'prop-types';

const convertArray = array =>
  array.map(type => convertPropType(type));

const convertShape = (shape) => {
  const result = {};
  Object.keys(shape).forEach((key) => {
    result[key] = convertPropType(shape[key]);
  });
  return result;
};

const convertPropType = (propType) => {
  let result;
  if (propType && propType.type) {
    if (!PropTypes[propType.type]) {
      throw new Error(`react-desc: unknown type ${propType.type}`);
    }
    if (propType.args) {
      if (propType.type === 'oneOfType' || propType.type === 'arrayOf') {
        if (Array.isArray(propType.args)) {
          result = PropTypes[propType.type](convertArray(propType.args));
        } else {
          result = PropTypes[propType.type](convertPropType(propType.args));
        }
      } else if (propType.type === 'shape') {
        result = PropTypes[propType.type](convertShape(propType.args));
      } else {
        result = PropTypes[propType.type](propType.args);
      }
    } else {
      result = PropTypes[propType.type];
    }
  } else {
    result = propType;
  }
  return result;
};

const decoratePropType = (propType, description, options = {}) => {
  if (!description) {
    throw new Error('react-desc: description is required');
  }
  if (!propType) {
    throw new Error('react-desc: propType is required');
  }

  let result = convertPropType(propType);
  if (propType.required === true || options.required) {
    result = result.isRequired;
  }
  result = result.bind(null);
  result.$$reactDesc = { description, propType, options };
  return result;
};

export default function decorate(component, metadata = {}) {
  if (!component) {
    throw new Error('react-desc: component is required');
  }
  if (!metadata.description) {
    throw new Error('react-desc: description is required');
  }

  const reactDesc = { ...metadata };
  delete reactDesc.props;
  const defaultProps = {};
  const propTypes = {};

  if (metadata.props) {
    Object.keys(metadata.props).forEach((name) => {
      const prop = metadata.props[name];
      // if it is an array it means that it has a description
      if (Array.isArray(prop)) {
        const [propType, description, options] = prop;

        if (options && options.defaultProp) {
          defaultProps[name] = options.defaultProp;
          delete options.defaultProp;
        }

        propTypes[name] = decoratePropType(propType, description, options);
      } else {
        propTypes[name] = prop;
      }
    });
  }

  /* eslint-disable no-param-reassign */
  component.$$reactDesc = reactDesc;
  if (!component.propTypes) {
    component.propTypes = propTypes;
  }

  if (!component.defaultProps) {
    component.defaultProps = defaultProps;
  }
  /* eslint-enable no-param-reassign */
}
