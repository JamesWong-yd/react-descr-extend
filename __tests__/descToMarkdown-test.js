import React from 'react';
import ReactPropTypes from 'prop-types';
import descToMarkdown from '../src/descToMarkdown';
import decorate from '../src/decorate';
import PropTypes from '../src/PropTypes';

class FakeComponent {}
const complexShape = {
  color: PropTypes.string,
  fontSize: PropTypes.number,
  one: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.instanceOf(FakeComponent),
    (props, propName) => typeof props[propName] === 'number',
  ]),
  test: PropTypes.instanceOf(FakeComponent),
  test2: PropTypes.shape({
    test3: PropTypes.string,
    test4: PropTypes.shape({
      test5: PropTypes.string,
      test6: PropTypes.number,
      test7: (props, propName) => typeof props[propName] === 'number',
      test8: PropTypes.arrayOf(PropTypes.string),
    }),
  }),
};

const DocumentedComponent = () => <div>fake</div>;

const DeprecatedComponent = () => <div>deprecated</div>;
DeprecatedComponent.propTypes = {
  testDeprecated: ReactPropTypes.string,
};
const NoPropTypeComponent = () => <div>fake</div>;

class ExtraInfoComponent {
  render() {
    return (
      <div>fake</div>
    );
  }
}

decorate(DocumentedComponent, {
  description: 'component',
  props: {
    test: [PropTypes.any, 'any'],
    test2: [PropTypes.array, 'array'],
    test3: [
      PropTypes.arrayOf(PropTypes.number), 'arrayOf',
    ],
    test4: [PropTypes.bool, 'bool'],
    test5: [() => {}, 'custom', {
      format: 'XXX-XX-XXXX',
    }],
    test6: [PropTypes.element, 'element'],
    test7: [PropTypes.func, 'func'],
    test8: [PropTypes.instanceOf(FakeComponent), 'instanceOf'],
    test9: [PropTypes.node, 'node'],
    test10: [PropTypes.number, 'number'],
    test11: [PropTypes.object, 'object'],
    test12: [PropTypes.objectOf(PropTypes.number), 'objectOf'],
    test13: [PropTypes.oneOf(['News', 'Photos']), 'oneOf'],
    test14: [PropTypes.oneOfType([
      PropTypes.number, PropTypes.string,
    ]), 'oneOfType'],
    test15: [PropTypes.symbol, 'symbol'],
    test16: [PropTypes.shape(complexShape), 'shape'],
    testRequired: [PropTypes.string, 'testRequired', {
      required: true,
    }],
    testDeprecated: [PropTypes.string, 'testDeprecated', {
      deprecated: 'use something else',
    }],
    testExtra: [PropTypes.string, 'testExtra', {
      defaultProp: 'abc',
      usage: 'test5 = "abc"',
    }],
  },
});
decorate(NoPropTypeComponent, {
  description: 'noPropType',
});
decorate(DeprecatedComponent, {
  description: 'component',
  deprecated: 'use button instead',
});
decorate(ExtraInfoComponent, {
  description: 'component',
  props: {
    test: [PropTypes.string, 'test', {
      defaultProp: 'abc',
    }],
  },
  usage: 'import component  from "grommet/components/Anchor"',
});

it('fails for missing component property', () => {
  expect(() => {
    descToMarkdown(undefined);
  }).toThrowError('react-desc: component is required');
});

it('documents empty doc for component without reactDesc', () => {
  const documentation = descToMarkdown(FakeComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents empty propType doc for component', () => {
  const documentation = descToMarkdown(NoPropTypeComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents a basic documented component', () => {
  const documentation = descToMarkdown(DocumentedComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents a deprecated documented component', () => {
  const documentation = descToMarkdown(DeprecatedComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents an extra info documented component', () => {
  const documentation = descToMarkdown(ExtraInfoComponent);
  expect(documentation).toMatchSnapshot();
});
