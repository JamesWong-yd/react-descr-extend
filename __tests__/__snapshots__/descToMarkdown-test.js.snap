// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documents a basic documented component 1`] = `
"## DocumentedComponent Component
component

### Properties

| Property | Description | Format | Default Value | Required | Details |
| ---- | ---- | ---- | ---- | ---- | ---- |
| **test** | any | any |  | No |  |
| **test2** | array | array |  | No |  |
| **test3** | arrayOf | [number] |  | No |  |
| **test4** | bool | boolean |  | No |  |
| **test5** | custom | XXX-XX-XXXX |  | No |  |
| **test6** | element | element |  | No |  |
| **test7** | func | function |  | No |  |
| **test8** | instanceOf | new FakeComponent(...) |  | No |  |
| **test9** | node | node |  | No |  |
| **test10** | number | number |  | No |  |
| **test11** | object | object |  | No |  |
| **test12** | objectOf | { test: number, ... } |  | No |  |
| **test13** | oneOf | News<br/>Photos |  | No |  |
| **test14** | oneOfType | number<br/>string |  | No |  |
| **test15** | symbol | symbol |  | No |  |
| **test16** | shape | {<br/>  color: string,<br/>  fontSize: number,<br/>  one: <br/>    string<br/>    number<br/>    new FakeComponent(...)<br/>    custom,<br/>  test: new FakeComponent(...),<br/>  test2: <br/>    {<br/>      test3: string,<br/>      test4: <br/>        {<br/>          test5: string,<br/>          test6: number,<br/>          test7: custom,<br/>          test8: [string]<br/>        }<br/>    }<br/>} |  | No |  |
| **testRequired** | testRequired | string |  | Yes |  |
| **~~testDeprecated~~** | testDeprecated | string |  | No | **Deprecated**: use something else |
| **testExtra** | testExtra | string | abc | No |  |"
`;

exports[`documents a deprecated documented component 1`] = `
"## ~~DeprecatedComponent~~ Component
### Deprecated: use button instead
component"
`;

exports[`documents an extra info documented component 1`] = `
"## ExtraInfoComponent Component
component

### Properties

| Property | Description | Format | Default Value | Required | Details |
| ---- | ---- | ---- | ---- | ---- | ---- |
| **test** | test | string | abc | No |  |"
`;

exports[`documents empty doc for component without reactDesc 1`] = `"## FakeComponent Component"`;

exports[`documents empty propType doc for component 1`] = `
"## NoPropTypeComponent Component
noPropType"
`;
